pcap-dpdk.c

```diff
@@ -0,0 +1,742 @@
+ *      notice, this list of conditions and the following disclaimer.
+ *   2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/*
+Date: Dec 16, 2018
+
+Description:
+1. Pcap-dpdk provides libpcap the ability to use DPDK with the device name as dpdk:[portid], such as dpdk:0.
+2. DPDK is a set of libraries and drivers for fast packet processing. (https://www.dpdk.org/) 
+
+Limitations:
+1. By default enable_dpdk is no, unless you set inlcudes and lib dir
+by --with-dpdk-includes= --with-dpdk-libraries=
+2. Only support link libdpdk.so dynamicly, because the libdpdk.a will not work correctly.
+3. Only support read operation, and packet injection has not been supported yet.
+4. I have tested on DPDK v18.11.
+Usage:
+1. compile DPDK as shared library and install.(https://github.com/DPDK/dpdk.git)
+
+You shall modify the file $RTE_SDK/$RTE_TARGET/.config and set:
+CONFIG_RTE_BUILD_SHARED_LIB=y
+
+2. launch l2fwd that is one of DPDK examples correctly, and get device information.
+
+You shall learn how to bind nic with DPDK-compatible driver by $RTE_SDK/usertools/dpdk-devbind.py, such as igb_uio.
+And enable hugepages by dpdk-setup.sh
+
+Then launch the l2fwd with dynamic dirver support. For example:
+$RTE_SDK/examples/l2fwd/$RTE_TARGET/l2fwd -dlibrte_pmd_e1000.so -dlibrte_pmd_ixgbe.so -dlibrte_mempool_ring.so -- -p 0x1
+
+3. compile libpcap with dpdk options.
+
+you shall run the following command to generate a new configure
+
+make clean
+autoreconf
+
+Then, run configure with dpdk options.
+For Ubuntu, they are --with-dpdk-includes=/usr/local/include/dpdk/ --with-dpdk-libraries=/usr/local/lib
+
+4. link your own program with libpcap, and use DPDK with the device name as dpdk[portid], such as dpdk:0.
+And you shall set DPDK configure options by environment variable DPDK_CFG
+For example, the testprogs/capturetest could be lanched by: 
+
+env DPDK_CFG="--log-level=debug -l0 -dlibrte_pmd_e1000.so -dlibrte_pmd_ixgbe.so -dlibrte_mempool_ring.so" ./capturetest -i dpdk:0
```

Comment:
So why do *the names of some shared objects* have to be part of the configure options?

Reply:
It's a long story.
1. Currently, we can not use static link with DPDK in libpcap. only *dynamic link* works correctly.

Becuase, DPDK uses global static variable to initialize NIC device. When DPDK is static linked by other program, this variable is NULL. As a result, the device can not be found when rte_eal_init() is called. More information is at: https://github.com/DPDK/dpdk/blob/master/drivers/bus/pci/pci_common.c lineno:548 . I tried `--whole-archive` flags, but DPDK provides many .a files not only one. And some .a files are mutually exclusive.

 2. So dynamic link is the only choice. But DPDK only provides runtime option `-d` to load device related plugins.

For example:
librte_pmd_e1000.so is for e1000 driver and is usually used for 1G ethernet NICs.
librte_pmd_ixgbe.so is for igxbe driver and is usually used for 10G ethernet NICs.
librte_mempool_ring.so is used for mempool.
More information is at: http://mails.dpdk.org/archives/dev/2015-May/017789.html

Finally, I designed this way to configure DPDK correctly.
I will provide two most common plugin default, but user can overwrite env variable DPDK_CFG, to launch their NIC related driver plugins.